### PR TITLE
DRILL-5808: Reduce memory allocator strictness for "managed" operators

### PR DESCRIPTION
**Pull request by paul-rogers**
The "managed" external sort and the hash agg operators now actively attempt to stay within a memory "budget." 

Out goals are to:

1. Stay within the budget, and
2. Make full use of the available memory.

Unfortunately, at present, Drill has a number of limitations that work at cross-purposes to the above goal.

* Upstream operators create record batches potentially larger than the memory budget.
* Memory allocations are "lumpy" - power of two rounded.
* Vectors double in size automatically when needed.

The combination of the above means that memory planning must be aware of the size of each and every vector to the byte level in order to predict size doubling and power-of-two rounding.

But, of course, Drill is schema-on-read, meaning that Drill cannot know ahead of time the "shape" of the data it will process. Without that information, memory estimates are, at best, averages, but actual allocations have a wide variance around those averages.

Add to this Drill's memory allocation scheme: each operator is given a strict budget enforced by the memory allocator. Go above the budget by a single byte and the query dies.

How do we resolve this conflict? On the one hand, Drill's internals are rough-and-ready; it is impossible to predict actual memory usage. On the other hand, the allocator requires perfect prediction else the user suffers with failed queries.

Much work is needed in Drill internals to provide for better memory management. (Relational databases have long ago solved the issues, so solutions are available.) Until then, this commit introduces a work-around.

Memory-managed operators can ask for "leniency" from the allocator. In this mode, the allocator:

* Allows actual memory use to spike up to 100% of the limit, or 100 MB, whichever is less,
* Logs each such "excess allocation" as a warning, so we can identify and fix issues, and
* Allows leniency only in production environments, but not during development or test.

That is, we give users a margin for error so that their queries succeed even if Drill's memory calculations don't come out exactly right.

This should be fine because, of course, Drill still has several operators that observe no memory limits at all. Seems silly to have one operator using GBs of memory, while enforcing a typical 30 MB limit on others.

The margins assume that the error in the calcs is normally distributed. For each operator (e.g. minor fragment) that goes over, another is under the limit, meaning that, on average, the aggregate of the set of operators is more-or-less at the limit.

Until all operators are memory managed, and Drill provides better memory management tools, this PR allows queries to succeed even if we get things slightly wrong internally.

### CODE REVIEW

File: exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java [hunk 1]
```diff
@@ -80,6 +119,23 @@ public Accountant(Accountant parent, long reservation, long maxAllocation) {
   }
 
   /**
+   * Request lenient allocations: allows exceeding the allocation limit
+   * by the configured grace amount. The request is granted only if strict
+   * limits are not required.
+   *
+   * @param enable
+   */
+  public boolean setLenient() {
```
Comment (Ben-Zvi) [inline exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java]: Why "set" returns a boolean ? Should better have a separate "get".

File: exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java [hunk 2]
```diff
@@ -22,16 +22,55 @@
...
+  // that are trying to stay within the limit, but are hindered by the
+  // current power-of-two allocations and random vector doublings. Instead
+  // of failing the query, the accountant allows an excess allocation within
+  // the grace, but issues a warning to the log.
+  //
+  // Grace is allowed only when requested, and then only in production
+  // code, or if enabled in debug code by setting the following system
+  // property.
+
+  public static final String STRICT_ALLOCATOR = "drill.memory.strict.allocator";
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java]: The tricky-bit is that the allocator/accountant is below the level of java-exec, and so does not have visibility to either Drill config or the option manager. So, this has to be set via a system option, as in the other memory manager controls.

A possible enhancement would be to grab the options in the Drill server startup (or main) and set statics down in the memory layer.

File: exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java [hunk 3]
```diff
@@ -22,16 +22,55 @@
...
+
+  // In large-memory allocators, don't allow more than 100 MB
+  // grace.
+
+  public static final int MAX_GRACE = 100 * 1024 * 1024;
+
+  // Use the strictness set in a system property. If not set, then
+  // default to strict in debug mode, lenient in production mode.
+
+  public static final boolean STRICT = System.getProperty(STRICT_ALLOCATOR) == null
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java]: This is a constant, hence the caps. It may be a bit confusing since it is set as an expression, but that is perfectly valid Java.

File: exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java (hunk 1 above)
Comment (paul-rogers) [reply exec/memory/base/src/main/java/org/apache/drill/exec/memory/Accountant.java]: Returned for testing, which is the only place that really cares about whether the leniency request was granted. I should, however, add a Javadoc comment to that effect.

**Comment by paul-rogers** [comment]
@Ben-Zvi, you are right that, in the worst case, this change will allow operators to exceed the memory allotment. But, that is actually the purpose.

As we know, it is *very* difficult to get memory management just right at present due to the wildly varying memory layouts for vectors, power-of-two rounding of buffer sizes, unexpected doubling of vectors, and lack of control over the size of incoming batches. We'd love to fix these, but doing so will take time.

In the meanwhile, we have the choice of failing queries because the calcs are off by a bit, or being more flexible and letting queries succeed at the risk of running out of memory. The change here does log each "excess" allocation so we can find them and fix any remaining issues. Also, in a test environment, strict limits are enforced to find bugs.

All of this is set against the backdrop of the exchange operators, hash join, and other operators that have an unlimited appetite for memory. Until we reign in those operators, seems silly to kill user queries because those operators that *do* manage memory make a small mistake here or there.

Once all operators are under control, and Drill's internal memory allocation is under better control, we can back out this change and be much more strict about enforcing memory limits.

Bottom line: should we fail user queries because of remaining rough spots in the "managed" operators? Or, should we allow user queries to succeed at a very small additional risk of running out of memory?
